cdc/sinkv2/eventsink/txn/mysql/mysql.go

```diff
@@ -0,0 +1,728 @@
+	}
+
+	// dsn format of the driver:
+	// [username[:password]@][protocol[(address)]]/dbname[?param1=value1&...&paramN=valueN]
+	username := sinkURI.User.Username()
+	if username == "" {
+		username = "root"
+	}
+	password, _ := sinkURI.User.Password()
+	hostName := sinkURI.Hostname()
+	port := sinkURI.Port()
+	if port == "" {
+		port = "4000"
+	}
+
+	// This will handle the IPv6 address format.
+	host := net.JoinHostPort(hostName, port)
+	dsnStr := fmt.Sprintf("%s:%s@tcp(%s)/%s", username, password, host, params.tls)
+	dsn, err := dmysql.ParseDSN(dsnStr)
+	if err != nil {
+		return nil, cerror.WrapError(cerror.ErrMySQLInvalidConfig, err)
+	}
+
+	// create test db used for parameter detection
+	// Refer https://github.com/go-sql-driver/mysql#parameters
+	if dsn.Params == nil {
+		dsn.Params = make(map[string]string, 1)
+	}
+	if params.timezone != "" {
+		dsn.Params["time_zone"] = params.timezone
+	}
+	dsn.Params["readTimeout"] = params.readTimeout
+	dsn.Params["writeTimeout"] = params.writeTimeout
+	dsn.Params["timeout"] = params.dialTimeout
+	testDB, err := GetDBConnImpl(ctx, dsn.FormatDSN())
+	if err != nil {
+		return nil, err
+	}
+	defer testDB.Close()
+
+	// Adjust sql_mode for compatibility.
+	dsn.Params["sql_mode"], err = querySQLMode(ctx, testDB)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	dsn.Params["sql_mode"], err = dmutils.AdjustSQLModeCompatible(dsn.Params["sql_mode"])
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	// NOTE: quote the string is necessary to avoid ambiguities.
+	dsn.Params["sql_mode"] = strconv.Quote(dsn.Params["sql_mode"])
+
+	dsnStr, err = generateDSNByParams(ctx, dsn, params, testDB)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	// check if GBK charset is supported by downstream
+	gbkSupported, err := checkCharsetSupport(ctx, testDB, charset.CharsetGBK)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if !gbkSupported {
+		log.Warn("gbk charset is not supported by downstream, "+
+			"some types of DDL may fail to be executed",
+			zap.String("hostname", hostName), zap.String("port", port))
+	}
+
+	db, err := GetDBConnImpl(ctx, dsnStr)
+	if err != nil {
+		return nil, err
+	}
+	db.SetMaxIdleConns(params.workerCount)
+	db.SetMaxOpenConns(params.workerCount)
+
+	ctx, cancel := context.WithCancel(ctx)
+	sink := &mysqlSink{
+		changefeedID:   changefeedID,
+		db:             db,
+		params:         params,
+		forceReplicate: replicaConfig.ForceReplicate,
+		enableOldValue: replicaConfig.EnableOldValue,
+		statistics:     metrics.NewStatistics(ctx, metrics.SinkTypeDB),
+		cancel:         cancel,
+	}
+
+	log.Info("mysql backend is created",
+		zap.String("changefeedID", fmt.Sprintf("%s.%s", changefeedID.Namespace, changefeedID.ID)),
+		zap.Bool("forceReplicate", sink.forceReplicate),
+		zap.Bool("enableOldValue", sink.enableOldValue))
+	return sink, nil
+}
+
+func (s *mysqlSink) OnTxnEvent(event *eventsink.TxnCallbackableEvent) (needFlush bool) {
+	if event.Event == nil {
```

Comment:
When can this happen? Please add comment on the contract of this method.

Reply:
Have commented on interface `backend`.

Reply:
why is the event.Event equal to nil ?